KAFKA-2390; OffsetOutOfRangeException should contain the Offset and Partition info.

lindong28 opened the pull request.

lindong28 committed:
KAFKA-2390; Seek() should take a callback

lindong28 committed:
fix checkstyle issue

hachikuji commented:
@lindong28 It looks like the callback is discarded without completion on a rebalance. I think we should probably make sure it always gets invoked eventually. A couple other points: 
- Do you think it would be helpful to include the TopicPartition and the offset that was seeked to in the callback?
- Should we provide a similar mechanism for seekToBeginning and seekToEnd? It could be a helpful way to find the new offset.

lindong28 commented:
@hachikuji Thanks for your quick review.

1) Currently the only use case we envision for this callback in seek() is when the position to seek is out of range. Since a rebalance will change the offset and invalidate the previous seek, I think we can invalidate the callback as well.

2) Yes, on a double think I think it will be useful to include the TopicPartition and the offset that was seeked to in the callback. This will be useful upon OFFSET_OUT_OF_RANGE exception.

3) Due to the use case of this callback, I personally don't think it is useful to include it in seekToBeginning and seekToEnd. I would also like to see what others think of this.

ewencp commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -407,7 +407,7 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
                 } else if (partition.errorCode == Errors.NONE.code()) {
                     long fetchOffset = request.fetchData().get(tp).offset;
-
+                    subscriptions.invokeConsumerSeekCallback(tp, null);
```
Shouldn't this be invoked only after we've checked that this consume request matches the expected offset? I think the current code will behave incorrectly given the following sequence of events:
1. Fetch request sent
2. Rebalance
3. Seek
4. Fetch response triggers this code

The callback is invoked, but the fetch request was for the previous generation and errors like OFFSET_OUT_OF_RANGE will not be correctly passed to the callback.

lindong28 replied:
@ewencp Thanks for the review! I will fix it in the patch.

ewencp commented:
I agree with @hachikuji that the callback should always be invoked. If it isn't, then I think there would need to be some _very_ clear javadocs indicating when it is/isn't invoked. Any use of this callback also requires implementing a ConsumerRebalanceCallback if you hold on to any state while waiting for the callback.

lindong28 committed:
Always invoke ConsumerSeekCallback; only invoke the callback if fetched offset matches the seeked offset

lindong28 commented:
@ewencp @hachikuji. Thanks for the suggestion! I have updated the patch such that the ConumserSeekCallback is always invoked, including when SubscriptionState.clearAssignment() and SubscriptionState.clearPartition(). I am not sure why the use of this callback also requires  implementing a ConsumerRebalanceCallback though. Let me double think on this.

ewencp commented:
@lindong28 I only meant it required implementing ConsumerRebalanceCallback if it _didn't_ get invoked in some cases. If the application did something like invoke the commit and wait for the callback, meanwhile holding on to resources, but the callback wasn't guaranteed to be invoked if a rebalance occurred in the meantime, then the only way to correctly clean up the resources associated with the callback would be to implement a ConsumerRebalanceCallback to clean them up.

If we guarantee the callback is always invoked, then this is no longer necessary (and simplifies application code that uses this callback).

hachikuji commented:
@lindong28 @ewencp I think the thing that's bugging me about the callback is that the semantics seem strange. When the consumer invokes seek(), the position is updated immediately and position() will return the seeked position. It's only later that the user finds out if that position is out of range. Intuitively, I would expect the callback to be invoked when the position is updated. I wonder if instead of using the generic ConsumerSeekCallback, the behavior would be clearer with an OffsetOutOfRangeCallback, which is only invoked when that error occurs? Then we wouldn't have to worry about ensuring the callback is invoked in all cases and it would make sense for the position to be updated immediately. Does that make sense?

ewencp commented:
@hachikuji Definitely understand that concern, I was confused at first that a callback even made sense since this seemed like an immediate update regardless of the mode the consumer was in. But would this OffsetOutOfRangeCallback be similar to the ConsumerRebalanceCallback where you need to provided it when you subscribe? It seems like it has the same somewhat weird callback semantics where it is provided once and can be invoked many times.

But now that I'm thinking about more of these types of events, is it possible we just have the name wrong? Maybe we should be implementing a `ConsumerListener` instead of a variety of callbacks? I think that naming is more in line with how we would normally describe the functionality since each of these methods can be invoked multiple times (rather than being a callback associated with a single invocation of a method).

In any case, for this specifically, it is a bit odd that the callback would be disconnected from the request, but on the other hand, shouldn't this type of event only ever occur during a seek? The only other time it happens is during initial subscription, right?

hachikuji commented:
@ewencp Usually when we get the offset out of range error, we just reset the position, so i wouldn't see that callback as one which could be invoked multiple times. The callback just provides a way to override the default reset behavior. It's also intuitive in my opinion to be able to handle the error at the same location in the code that the seek is performed.

I agree, however, that this is still not great. The user might want to intercept all out of range errors, not just the ones that result from a specific seek. The `ConsumerListener` interface could be interesting (at the very least, I definitely would approve of changing `ConsumerRebalanceCallback` to `ConsumerRebalanceListener`). It might be worthwhile to spend a little time trying to flesh out a basic sketch of the interface for comparison.

lindong28 commented:
@ewencp @hachikuji I think we are discussing to things: 1) whether it is more intuitive to call it ConsumerSeekCallback or OffsetOutOfRangeCallback; 2) whether we should turn it and other callback into listener instead. I agree with your opinion on 2). But 2) goes beyond the scope of this ticket and requires much detailed discussion since it changes the user interface (e.g. ConsumerRebalanceListener). Can we focus on 1) first?

I would prefer ConsumerSeekCallback to OffsetOutOfRangeCallback. If the user expects ConsumerSeekCallback to be invoked when position is updated, shouldn't he has the same expectation for OffsetOutOfRangeCallback? By keeping ConsumerSeekCallback, we have a more general callback API, and we can ensure that it is always invoked, including when rebalance happens or when user re-subscribe to partitions. We can describe in the document that seek() is actually an asynchronous API and it is only effective when the fetch after the seek gets a non-error response.

hachikuji commented:
@lindong28 That's a fair point about OffsetOutOfRangeCallback. It also has confusing semantics.

I think what @ewencp was suggesting with the `ConsumerListener` interface is basically another approach to the problem that the `ConsumerSeekCallback` is trying to address. We all agree that the user needs some way to be able to handle out of range errors. The weird thing with the seek callback is that it only allows the user to hook into out of range errors when they seek, but these errors are an ongoing concern during the lifetime of the consumer. You might seek, fetch successfully, and then get an out of range error on the next fetch. The seek callback doesn't help you in that case, so you have to resort to catching exceptions. The question is whether we would need `ConsumerSeekCallback` any more if we had a `ConsumerListener` which provided a general hook for out of range errors. The nice thing is that it bypasses the weirdness of when the position gets updated and when the seek callback gets invoked. What do you think?

lindong28 commented:
@hachikuji I agree that a ConsumerListener which provides a general hook for out of range errors is useful. What I try to understand is, whether there is a valid user case to have a callback for a specific seek. I just discussed with @becketqin and tend to think it can still be useful to provide callback in seek().

An out-of-range exception can happen for one of the three reasons: 1) invalid offset in seek, 2) unclean leader election, and 3) consumer lags behind too much. One reasonable strategy that I envision is that, consumer will throw exception and quit for 1), reset offset to latest for 2), and reset offset to earliest for 3). Such a strategy can be implemented using a combination of ConsumerSeekCallback and ConsumerListener, but not with ConsumerListener alone. Does this usecase make sense?

hachikuji commented:
@lindong28 Fair enough. I do see some value in being able to handle an out-of-range error in the scope of a specific seek. I still think the behavior around when the position gets updated is a little weird, but perhaps documenting it clearly will be enough. We have to explain how it is synchronous in the sense that the position will be updated immediately and the next fetch will try to use it, but asynchronous in the sense that we can only find out if the position is valid after we try to fetch. It makes it a little tough to say when the seek is "complete." Maybe instead of using `onComplete` in the callback, we could rename it to something which reflects this difference? Maybe `onValidated`?

It would be cool if there was an API on the broker to get the current range of a partition. Then we could have this method block while we check that the offset is in range. You could do it with the list offset request, but it seems like it would require two requests, which is annoying.

lindong28 commented:
@hachikuji Of course, thanks for your suggestion. I will clarify it in document and use onValidated.

lindong28 committed:
Change callback name and API document of ConsumerSeekCallback to better explain its usage

lindong28 commented:
@hachikuji Sorry for the delay. I just updated patch to rename the callback to onResponse and updated API documentation to illustrate the reason that the callback is asynchronous. Does that alleviate the concern you have?

I thought onResonse is a little better than onValidated because it hints the time when this callback is called. And the name is more general rather than being tied to a specific functionality, e.g. validation. What do you think?

Thank you!

hachikuji commented:
@lindong28 Yeah, I didn't like onValidated that much either. onResponse seems a little better. How about onFetchResponse to be even more explicit? Or maybe onFirstFetch to emphasize that the callback only applies to the first fetch? It's a little tough to tell how intuitive any of these will be for users.

lindong28 committed:
use a more intuitive name for ConsumerSeekCallback function

lindong28 commented:
@hachikuji Sure. These names are better. I combined your suggestions and named it as onFirstFetchReponse. What do you think?

lindong28 committed:
undo all changes made earlier

lindong28 committed:
Merge branch 'trunk' into KAFKA-2390

lindong28 committed:
KAFKA-2390; OffsetOutOfRangeException should contain the Offset and Partition info.

lindong28 commented:
@becketqin @onurkaraman Could you review the updated patch if you have time? Thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -256,12 +261,19 @@ private long listOffset(TopicPartition partition, long timestamp) {
+     * @throws OffsetOutOfRangeException If there is OffsetOutOfRange error in fetchResponse
      */
-    public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecords() {
+    public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecordsOrException() {
```
Is it really necessary to rename this method? The javadoc seems clear enough.

lindong28 replied:
Thanks for your review. I don't have strong preference on this name.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
```
I'm not really clear why this is done in a separate loop. Are you trying to avoid the work of collecting valid records unnecessarily? Also, I think the loop below still has the OFFSET_OUT_OF_RANGE check.

lindong28 replied:
Because I need to avoid making any change to subscriptions if there is OffsetOutOfRange error.

You are right. Loop below shouldn't need OFFSET_OUT_OF_RANGE any more. I will remove it. Thanks for the catch.

hachikuji replied:
Why not let the subscriptions be updated? After all, the records that were successfully fetched are still the ones we want. We can add them to the records collection and they will be retrieved on the subsequent call to fetchedRecords() (the first call will throw the exception). Do you see any reason why that wouldn't work?

lindong28 replied:
Actually I have thought of this. I agree that, with proper implementation we can use the partition data that doesn't avoid error so that we save bandwidth. But I am not doing it because it is harder to reason the code when your records returned from poll() may contains data from "multiple" fetches -- if you have reassignment you may have reordered data as well.

It is certainly possible to think through all the scenario to make it right. But this complicates both this and future implementation. It doesn't seem worth it because the OffsetOutofRange exception should be pretty rare and the bandwidth used to re-fetch data should be very small.

hachikuji replied:
I'm not sure that this is such a big problem. The records accumulate until they are retrieved and we always check that the record offset returned to the user is the one they expect. It seems like the worst thing that could happen is that the records continue to grow while we keep throwing out of range exceptions to the user. To address this, we should probably have KafkaConsumer.poll() check whether there are records to return before calling initFetches(). I think that would guarantee that no new fetches would be sent until all pending data is returned to the user. Do you think that would work?

lindong28 replied:
Then we may have the following sequence of events:

1) consumer.poll() successfully returns data for partition 1 and 2
2) consumer.seek() gives partition 1 an invalid position
3) consumer.poll() throws OffsetOutOfRange exception for partition 1
4) consumer.seek() to another invalid position for partition 1
5) consumer.poll() returns data for partition 2
6) consumer.poll() throws OffsetOutOfRange exception for partition 1
...

It appears a bit counter intuitive to me that the 2nd poll() after the seek() caused the OffsetOutOfRange exception. In short it adds a little bit code complexity for a little bit performance improvement. I still slightly prefer the current solution though. I will do it if you are strong on your solution.

lindong28 replied:
I think I am preferring the existing solution more -- it will be easier to say that poll() will always request data for all partitions starting from the offset before poll() is called, and it either throw exception, or return data for all partitions that have new data after the current offset.

The proposal you suggested makes the API harder to explain -- it requires user to know whether there is an OffsetOutOfRange exception in the previous poll() call.

hachikuji replied:
It's definitely not a strong preference and the performance aspect doesn't really concern me either way. I am just trying to think through the full implications. One implication of discarding fetches is that it would be impossible to make progress unless the user handles the exception (either by propagating it or seeking to another offset). It's not clear to me if this is desirable or not. I also wonder whether some of these seek cases you mention are handled currently. Wouldn't it be possible even with the current patch to do a seek while a fetch is pending? In that case, the out of range error would apply to the previous offset and not the current one. That suggests to me that we ought to be checking the consumed position in SubscriptionState before throwing the exception.

hachikuji replied:
Also, I'm not really clear why the API would be any more difficult to explain if we keep the fetches. Any call to poll() can return data from the current consumed positions. What difference does it make when the data was actually fetched?

lindong28 replied:
I think the API for poll() should **ideally** say the following:

if (fetch offset for any partition is out-of-range)
  throw OffsetOutOfRange exception
else {
  return data for all partitions who have data after the current fetch offset, before timeout
}

The user can easily know whether the offset is valid and whether there is data by looking at vale returned from poll().

I think the API would be more difficult to explain with your proposal, because I thought it would become the following:

if (there is OffsetOutOfRange exception for other partition in previous patch && 
    there is data for this partition from previous patch)
  return data
else if (fetch offset for any partition is out-of-range)
  throw OffsetOutOfRange exception
else {
  return data for all partitions who have data after the current fetch offset, before timeout
}

In this case, suppose user receives data for partition x from poll(), he can no longer decide if it is because the current offset is valid.

However, my concern is invalid, since value returned from poll() doesn't necessary come from the most recent fetchRequest.

Neither of us are strong on either solutions. I will update the patch as you suggested and move forward:)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
+                    offsetOutOfRangePartitions.put(tp, fetchOffset);
+            }
+            if (!offsetOutOfRangePartitions.isEmpty())
+                this.exceptionFromLastFetch = new OffsetOutOfRangeException(offsetOutOfRangePartitions);
```
Maybe instead of storing the exception as a field, you can store the map and create the exception in fetchedRecords() if the map is not empty?

lindong28 replied:
I think both work well. But I slightly prefer the current approach. If the only reason we want to store the map is because we will create the exception afterwards, why not do it in the first place? In the future we might want to store other exception as well. What do you think?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -256,12 +261,19 @@ private long listOffset(TopicPartition partition, long timestamp) {
-    public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecords() {
+    public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecordsOrException() {
         if (this.subscriptions.partitionAssignmentNeeded()) {
             return Collections.emptyMap();
```
Should we clear the exception here? If not, then we'll have the exception thrown after the rebalance.

lindong28 replied:
I think we should not clear exception, because it should be handled in the same way we handle Fetcher.records -- we will still return already fetched records, instead of clearing it, after the rebalance. What do you think?

hachikuji replied:
Hmm... In that case, we should check that the out of range partitions are still assigned when the exception is thrown, right?

lindong28 replied:
Ah, good point! I missed it. Actually this is a good reason why we should store map first and create exception later. Let me update the patch.

lindong28 committed:
throw exception only if defaultResetStrategy != None

lindong28 commented:
@hachikuji Here is summary of changes:

1) Reverted the renaming of function fetchedRecords()
2) Removed no-longer-needed OFFSET_OUT_OF_RANGE in handleFetchResponse()
3) Only throw exception if defaultResetStrategy != None.

Here is the change that you recommended but I haven't made:

1) Clear exception in fetchedRecords() if partitionAssignmentNeeded() == true
2) Store the map and create the exception in fetchedRecords() if the map is not empty

Please let me know if you have any comment on the updated patch. Thanks again for taking time to review the patch!

lindong28 committed:
only throw exception for partitions that are still fetchable

Commit:
Don't initiate fetch if we already have records available from previous fetchRequest

lindong28 committed:
fix test failure

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -256,12 +266,28 @@ private long listOffset(TopicPartition partition, long timestamp) {
+
+            // filter offsetOutOfRangePartitions to retain only the fetchable partitions
+            for (Map.Entry<TopicPartition, Long> entry: this.offsetOutOfRangePartitions.entrySet()) {
+                if (!subscriptions.isFetchable(entry.getKey())) {
```
I might be mistaken, but this doesn't seem sufficient. It seems like we should compare the out of range offset with the current consumed position and only throw the exception if they are equal. Presumably if the user has seeked to a new position, then they wouldn't be interested in an out of range error from a previous fetch.

lindong28 replied:
@hachikuji Good catch. I fixed it in the updated patch. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -256,12 +266,28 @@ private long listOffset(TopicPartition partition, long timestamp) {
+            }
+            this.offsetOutOfRangePartitions.clear();
+            if (!currentOutOfRangePartitions.isEmpty())
+                throw new OffsetOutOfRangeException(currentOutOfRangePartitions);
```
Could we turn this block into a method? For example, throwIfOutofRange() or something like that.

lindong28 replied:
Sure. Turned this into a method.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -400,12 +426,34 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
+                     * Don't read data if there is OffsetOutOfRangeException for any partition in the fetchResponse.
+                     * We will fetch the data for all partitions in the fetchRequest again.
+                     */
+                    if (!this.offsetOutOfRangePartitions.isEmpty())
```
I thought that retaining the fetches meant we wouldn't need this check anymore and we could merge the two loops.

lindong28 replied:
Thanks for catching these. Yesterday I just want to quickly finish this patch and I didn't look through the code again after addressing the last comment..

lindong28 committed:
fix code to address reviews

lindong28 committed:
minor fix

lindong28 commented:
Couldn't understand why it can not compile by the robot... I can compile it without problem.

hachikuji commented:
Seems to be complaining about this line:
```
private void throwIfOffsetOutOfRange() throws OffsetOutOfRangeExceptionc
```
Maybe the extra 'c'?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```
The only thing I'm puzzling over is the impact of this check. Currently, KafkaConsumer has the following snippet in poll():

``` java
fetcher.initFetches(cluster);
client.poll(timeout);
return fetcher.fetchedRecords();
```

I'm wondering in particular about that `client.poll(timeout)`. If we haven't actually initiated any new fetches, then there wouldn't be any expected IO and we'd block for the timeout duration unnecessarily. You could instead change it to something like this:

``` java
Map records = fetcher.fetchedRecords();
if (!records.isEmpty())
  return records;
fetcher.initFetches(cluster);
client.poll(timeout);
return fetcher.fetchedRecords();
```

Then we could remove this check and we wouldn't block unless there actually were pending fetches. What do you think?

lindong28 replied:
Nice catch! I think you are referring to `fetcher.initFetches(cluster)` in `pollOnce(timeout)` in `KafkaConsumer`, right?

I agree with you that there will be unnecessary expected IO. However, according to the API description of `pollOnce(timeout)`, it is expected to do one round of polling. Simply making the change you suggested would violate this API.

It is possible that we spend more time on thinking whether it is beneficial to make this API change. But this requires thorough thinking which is not a simple job. I also worry that this discussion will become longer and longer. 

Given the problem you just mentioned, how about I revert to my earlier approach, i.e. we discard already fetched data upon OffsetOutOfRange exception and fetch the entire data again. What do you think?

hachikuji replied:
I feel a little more inclined to try and make this API work since the implementation feels more natural, but I agree we should make sure we understand the impact and revert if we can't find a nice way to deal with it. Since `pollOnce` is an internal API, I'm not sure it actually has to invoke poll. At one point, I considered renaming it to `pollRecords` instead to emphasize that the main point was fetching data. If we did that, then I would see no inconsistency in shortcutting when records are already available. 

I think the thing that we're having a little trouble wrapping our heads around is the fact that a single fetch response can affect two invocations of `fetchedRecords`. Basically we're trying to deal with this by introducing another call to fetchedRecords. Another way we could handle it is to create a new method, for example, `Fetcher.outOfRangePartitions`, which would allow us to move the exception throw into KafkaConsumer and remove the need for the extra call to fetchedRecords. What do you think?

lindong28 replied:
I think we both agree that buffering the fetched data will require much more thought and effort than throwing away data and re-fetch. Can you elaborate the benefit of buffering fetched data over throwing away data, that makes these effort worthwhile?

Specifically, could you explain why it is more natural to buffer the data? I think throwing away data is more intuitive -- user calls poll(); there is exception; user handles it; and try poll() again. It is common, for example in batch processing system, to start your job all over after exception and failure handling. Starting from where you left is really an optimization that comes later in time.

hachikuji replied:
There were two things I didn't like about the discard approach. First was somewhat stylistic. Having two passes over responses to look at error conditions looks a little smelly. Second, the decision to throw away the data seems arbitrary. Why should we discard good data just because it happened to come in on a request with an unrelated error? And note that we only discard the data from the request that had an out of range error: fetches on other brokers can still succeed, which means there could still be data available in the record set at the same time that an out of range exception is waiting to be thrown. Now that I think about it again, you would probably need the additional call to fetchedRecords for that approach also, so I'm not sure discarding the data really buys you anything.

hachikuji replied:
Ah, I think I was wrong on one point. Your initial patch discarded ALL data until the exception was thrown. That would resolve the issue I mentioned above. That approach is easier to reason about if we could find a way to collapse the loops to remove some of its ugliness. We might also want to skip sending new fetches until the exception was handled to avoid fetching data we know will be discarded anyway. That would have implications for pollOnce (similar as above). Haha, not sure either approach is safe from complexity.

lindong28 replied:
Thanks for your comments. In order to keep data from ALL partitions of ALL `fetchResponse` which doesn't have `OffsetOutOfRangeException`, how about we do the following:

In the `initFetches()`, we only create `fetchRequest` for those partitions whose fetched position == consumed position. By doing so, `initFetch()` will only issue `fetchRequest` for partitions which caused `OffsetOutOfRangeException`; `poll(timeout)` will unblock upon `fetchResponse`; and the next successful `fetchedRecords()` will return data from all fetched partitions since the last successful `fetchedRecords()`.

Can you find anything problematic with this approach?

hachikuji replied:
Thanks for your patience with this issue, but I think it's really important that we think it all the way through. It actually occurs to me that one of the reasons this is turning out to be a little tricky is that the current code might not be handling all the cases around fetches correctly. I think the fact that `fetchedRecords` is not checked before blocking in poll() might already cause unnecessary blocking. The trouble is that the code seems to assume that fetches will only return in KafkaConsumer.poll(), but they could return in any blocking call (e.g. in a synchronous commit) since everything just delegates to NetworkClient.poll() in the end. That means that KafkaConsumer.poll() will block even when data is available. My proposal would probably be to leave Fetcher as it is in the current patch and add the check for non-empty records in KafkaConsumer.pollOnce() as I suggested before. Does that make sense?

lindong28 replied:
Thanks for your patience as well.

Talking about your proposal, I am not convinced that we should avoid calling `poll()` in `pollOnce()` when data is available, for the following reasons.

1) Can you elaborate why the fact that fetchedRecords is not checked before blocking in poll() might already cause unnecessary blocking? In other word, under which sequence of events would it cause unnecessary blocking?

2) If we add the check for non-empty records in KafkaConsumer.pollOnce(), this would cause internal API change. We have agreed that this is a problem, where you proposed to use `Fetcher.outOfRangePartitions`.

What you described may actually be a valid problem and worth effort to fix. But I am not strong on this at this moment. Creation of a new ticket involving API change would require the creator to clearly describe the motivation and use case. I would rather leave it to a new ticket, to be taken by you or other developers, unless there is no other way to finish this ticket.

I would appreciate it to first get opinion on whether my latest proposal can solve the this ticket. Thank you much for review.

hachikuji replied:
For 1), consider this sequence:

consumer.poll(100); // fetches sent here
consumer.commit(SYNC); // fetches returned here
consumer.poll(100); // should return immediately, but will block on the next round of fetches because we don't check for available records before polling

For 2), I'm a little unclear why you seem reluctant to change pollOnce. I'm pretty sure that the only place it's used is in KafkaConsumer.poll().

lindong28 replied:
For 1), I am more inlined rename or create a new call in KafkaConsumer, say fetchRecords(). Because the user of poll() might also expect it to do other IO, e.g. heartbeating, auto-commits, instead of only fetching data.

For 2), I am reluctant to make changes that I am not strong with. Because I am concerned that, if I do this change, another reviewer will question it which I can not defend. Then I need to discard the change I have made, which happened once in this ticket. I would be more confident to make this change if 1 committer or 2 senior non-committers can support it.

lindong28 replied:
BTW, your usecase also suggests that the change can be qualified as an API change. The ticket for this change would look like "Allow KafkaConsumer to avoid network IO when there is  data available to return".

hachikuji replied:
The change I'm suggesting is fully compatible with KafkaConsumer's poll() interface. It returns data when data is available. If the current behavior blocks with data available, then it is not following the API. Fixing this issue here is purely for convenience since it also resolves the same issue you have with unnecessary blocking in the current patch.

If we want to ensure that NetworkClient.poll is always invoked, you could change pollOnce to include this snippet instead:

``` java
Map records = fetcher.fetchedRecords();
if (!records.isEmpty()) {
  client.poll(0);
  return records;
} else {
  fetcher.initFetches(cluster);
  client.poll(timeout);
  return fetcher.fetchedRecords();
}
```

(By the way, I'm not a committer, but I did write pollOnce if that's worth anything.)

I don't have a strong preference for this solution, so if you can think of another, I encourage you to implement it. Probably the initial approach would get -1 from me since it makes the fetch response handling more convoluted. Have you talked with @onurkaraman and @becketqin to get more opinions?

lindong28 replied:
Yeah I have talked with both @onurkaraman and @becketqin. They thinks throwing away the fetched data is fine and simple. But this is a long thread and I haven't discussed other details with them.

BTW, I am not going to use initial approach, i.e. throw away data and have the loop. I am going to try the last proposal I described, which I copy & past below. This should keep the fetch response handling the same as it is in this patch.

In the `initFetches()`, we only create `fetchRequest` for those partitions whose fetched position == consumed position. By doing so, `initFetch()` will only issue `fetchRequest` for partitions which caused `OffsetOutOfRangeException`; `poll(timeout)` will unblock upon `fetchResponse`; and the next successful `fetchedRecords()` will return data from all fetched partitions since the last successful `fetchedRecords()`.

Though I think it belongs to a difference patch, I will apply the change you described above. Let's get it committed this altogether if it gets +1 from committer.

lindong28 replied:
BTW, calling `client.poll(0)` completely solves my concern I had earlier with API change. I think this is certainly an improvement.

lindong28 commented:
BTW calling client.poll(0) in pollOnce() completely solves my concern of API change. I think is a definitely an improvement.

lindong28 committed:
only create fetchRequest for partitions whose consumed == fetched offset

lindong28 committed:
remove unnecessary empty line and line movement

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -374,7 +406,9 @@ private void handleListOffsetResponse(TopicPartition topicPartition,
                 long offset = this.subscriptions.fetched(partition);
-                fetch.put(partition, new FetchRequest.PartitionData(offset, this.fetchSize));
+                long consumed = this.subscriptions.consumed(partition);
+                if (consumed == offset)
```
Might be worth a comment explaining why this check is needed. Basically it ensures that Fetcher will have no more than one outstanding request for any partition, which prevents fetches from getting too far ahead. Is that right? You might also mention that it doesn't impact prefetching since the consumed position will always be updated prior to initializing the next fetches.

Also, rename 'offset' to 'fetched'?

lindong28 replied:
Not exactly. This is actually a bug fix. This check is to make initFetch() consistent with handleFetchResponse() -- if consumed != fetched, then the fetched data will be discarded in handleFetchResponse() anyway. Even worse, handleFetchResponse() will reset fetched to consumed, re-fetch the already fetched data again, which is then discarded in fetchedRecords().

Sure, let me rename it to fetched.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -767,6 +767,11 @@ public void assign(List<TopicPartition> partitions) {
 
         // init any new fetches (won't resend pending fetches)
         Cluster cluster = this.metadata.fetch();
+        Map records = fetcher.fetchedRecords();
```
You'll probably have to add the generic parameters to the Map declaration.

hachikuji replied:
Also it'd be nice to have a comment basically saying that this check is needed since fetches may have already returned in another API call (e.g. commit).

lindong28 replied:
Sure, I will add the comment and generic parameters to the Map declaration.

hachikuji commented:
@lindong28 Left a few minor comments. Overall, LGTM.

lindong28 commented:
Thanks for all the comments @hachikuji ! Will update the patch soon.

lindong28 committed:
add comments

lindong28 commented:
"./gradlew testAll" can successfully finish on my desktop.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -80,6 +81,8 @@
     private final Deserializer<K> keyDeserializer;
     private final Deserializer<V> valueDeserializer;
 
+    private Map<TopicPartition, Long> offsetOutOfRangePartitions;
```
Should be final?

ijuma commented on clients/src/main/java/org/apache/kafka/common/errors/OffsetOutOfRangeException.java:
```diff
@@ -19,10 +22,15 @@
 public class OffsetOutOfRangeException extends RetriableException {
 
     private static final long serialVersionUID = 1L;
+    Map<TopicPartition, Long> offsetOutOfRangePartitions = null;
```
Should be private and final probably.

ijuma commented on clients/src/main/java/org/apache/kafka/common/errors/OffsetOutOfRangeException.java:
```diff
@@ -35,4 +43,8 @@ public OffsetOutOfRangeException(String message, Throwable cause) {
         super(message, cause);
     }
 
+    public Map<TopicPartition, Long> getOutOfRangePartitions() {
```
In Kafka, we tend to write the method name without the `get`, ie `outOfRangePartitions()`. The field name should have the same name though.

lindong28 replied:
@ijuma Thanks much for your review. Will pay attention to final, private and get method name in the future patch.

Please see updated patch. I didn't add final to variable offsetOutOfRangePartitions in class OffsetOutOfRangeException because otherwise I need to add 4 additional lines offsetOutOfRangePartitions=null to 4 constructor method. This seems redundant. Can you let me know if you prefer me adding final keyword and the additional 4 lines?

ijuma replied:
Thanks @lindong28.

Regarding the final field in `OffsetOutOfRangeException`, it's a good question as there is no nice solution. :) Ideally, we would not have 4 constructors since only 1 or 2 tend to be used, but too late to change that now. I think it's OK to leave it as you have as I don't know how committers feel about this particular issue (and I don't want to steer you towards a direction that they may disagre with!).

lindong28 replied:
Thanks @ijuma for your review!

lindong28 committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-2390

lindong28 committed:
add final and private keyword

lindong28 commented:
Hi @guozhangwang, this patch has been reviewed by @hachikuji and @ijuma. This patch has been pending for a long time. Since you are very familiar with new consumer code, can you have a look and see if it can be committed?

I am not sure if we are going to deliver new consumer in 0.9.0 release. If we do, then we probably need to include in 0.9.0 all patches that affect new consumer API, such as this patch and others under KAFKA-2387.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -258,15 +263,42 @@ private long listOffset(TopicPartition partition, long timestamp) {
+                log.debug("Ignoring fetched records for {} since it is no longer fetchable", entry.getKey());
+                continue;
+            }
+            Long consumed = subscriptions.consumed(entry.getKey());
```
Could you add some comments here for the checking of the consumed position?

guozhangwang commented:
@lindong28 LGTM overall, just one minor comment.

lindong28 committed:
add comment

lindong28 commented:
@guozhangwang Thanks! I have added comment. Please see updated patch.

hachikuji commented:
@lindong28 @guozhangwang We might need a followup on this patch. I think this line in throwIfOffsetOutOfRange doesn't work:

``` java
Long consumed = subscriptions.consumed(entry.getKey());
// ignore partition if its consumed offset != offset in fetchResponse, e.g. after seek()
if (consumed != null && entry.getValue() == consumed)
    currentOutOfRangePartitions.put(entry.getKey(), entry.getValue());
```

I think we need to use equals instead of '==':

``` java
Long consumed = subscriptions.consumed(entry.getKey());
// ignore partition if its consumed offset != offset in fetchResponse, e.g. after seek()
if (consumed != null && consumed.equals(entry.getValue()))
    currentOutOfRangePartitions.put(entry.getKey(), entry.getValue());
```

We should probably have a unit test for this case.

lindong28 commented:
@hachikuji Thanks for the catch! I have created a pull request to add unit test and replace == with equals.

Interestingly, I find that here == has the same effect as equals -- they can both update currentOutOfRangePartitions and allow unit test to pass. I don't have explanation though.

ijuma commented:
@lindong28 `==` is the same as `equals` when boxing values from -128 to 127 (or if you call `Long.valueOf`)

hachikuji commented:
@lindong28 This puzzled me for a little while as well, but it looks like the test case actually ends up using the same Long instance for both the fetch position (which is stored for out of range errors) and the consumed position when you use seek(), hence reference equality works fine. This won't generally be the case since the fetch position usually is updated independent of the consumed position.

lindong28 commented:
@ijuma Interesting. I am curious motivation for having this rule. Do you have any reference to related documentation or online discussion?

@hachikuji This puzzles me as well. Anyway, thanks much finding the problem.

ijuma commented:
@lindong28, it's because there's a cache for commonly used values for Integer, Long, etc. The cache exists to avoid excessive allocation and it was added in Java 5 before HotSpot supported escape analysis and scalar replacement. A reference that I quickly found via Google:

https://dzone.com/articles/surprising-results-autoboxing

lindong28 commented:
@ijuma Ah. So it called autoboxing. Thanks for this information:)
